### PR TITLE
feat(artifact): save to stream. Fixes #3189

### LINKED ISSUE #3189: optimization log
# Summary

Modify log save method to stream

# Motivation

Modify the log saving mode to streaming mode to prepare for the next real-time log saving

# Proposal
Save(path string, outputArtifact *wfv1.Artifact) error [link](https://github.com/argoproj/argo/blob/04c77f490b00ffc05f74a941f1c9ccf76a5bf789/workflow/artifacts/artifacts.go#L25)    to
 Save(stream Stream, outputArtifact *wfv1.Artifact) error 

current  :    docker  --->  local file(Disk) ----> Artifact
want :  docker----> local (RAM) ---->Artifact

### PR DESCRIPTION
**Pull request by wangzhipeng**
https://github.com/argoproj/argo/issues/3189  optimization log

### CODE REVIEW

File: workflow/artifacts/artifactory/artifactory.go [hunk 1]
```diff
@@ -73,3 +73,22 @@ func (a *ArtifactoryArtifactDriver) Save(path string, artifact *wfv1.Artifact) e
...
+	req.SetBasicAuth(a.Username, a.Password)
+	res, err := (&http.Client{}).Do(req)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		_ = res.Body.Close()
+	}()
+	if res.StatusCode < 200 || res.StatusCode >= 300 {
+		return errors.InternalErrorf("saving file to artifactory failed with reason:%s", res.Status)
```
Comment (alexec) [inline workflow/artifacts/artifactory/artifactory.go:91]: must support 404 and return correct error

File: workflow/artifacts/oss/oss.go [hunk 2]
```diff
@@ -75,3 +78,27 @@ func (ossDriver *OSSArtifactDriver) Save(path string, outputArtifact *wfv1.Artif
 		})
 	return err
 }
+
+func (ossDriver *OSSArtifactDriver) Write(reader io.Reader, outputArtifact *wfv1.Artifact) error {
+	err := wait.ExponentialBackoff(wait.Backoff{Duration: time.Second * 2, Factor: 2.0, Steps: 5, Jitter: 0.1},
```
Comment (alexec) [inline workflow/artifacts/oss/oss.go:83]: can I ask how you test OSS please?

**Comment by wangzhipeng** [comment]
Sorry, this is one of my ideas, I am doing the job of storing logs in real time to the remote end, let me close this PR first

**Comment by alexec** [comment]
I think this is a good idea.

**Pull request closed without merging**